artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQMessageProducer.java

```diff
@@ -533,6 +496,52 @@ private void doSendx(ActiveMQDestination destination,
       }
    }
 
+   private void checkDestination(ActiveMQDestination destination,
+                                 SimpleString address,
+                                 ClientSession clientSession) throws JMSException {
+      if (!connection.containsKnownDestination(address)) {
+         try {
+            ClientSession.AddressQuery query = clientSession.addressQuery(address);
+
+            if (!query.isExists()) {
+               checkQueue(destination, address, clientSession, query);
+            } else {
+               if (destination.isQueue()) {
+                  ClientSession.QueueQuery queueQuery = clientSession.queueQuery(address);
+                  if (!queueQuery.isExists()) {
+                     checkQueue(destination, address, clientSession, query);
+                  }
+               }
+
+               connection.addKnownDestination(address);
+            }
+         } catch (ActiveMQQueueExistsException e) {
+            // The queue was created by another client/admin between the query check and send create queue packet
+         } catch (ActiveMQException e) {
+            throw JMSExceptionHelper.convertFromActiveMQException(e);
+         }
+      }
+   }
+
+   private void checkQueue(ActiveMQDestination destination,
+                                SimpleString address,
+                                ClientSession clientSession,
+                                ClientSession.AddressQuery query) throws ActiveMQException, InvalidDestinationException {
+      if (destination.isQueue() && query.isAutoCreateQueues()) {
```

Comment:
should this not also, dare i say check: query.isAutoCreateAddresses()

Reply:
e.g. no point calling,          clientSession.createAddress(address, RoutingType.ANYCAST, true); 

if its now allowed....

Reply:
I agree with you. although It wasn't part of my original change.. I was just fixing a bug

Reply:
let me rewrite this